### PR TITLE
HPCC-21186 Service level transforms should run for all methods

### PR DESCRIPTION
**Pull request by afishbeck**
We should treat service level transforms as common to all methods.
We can provide a way to selectively call common transforms in the
future.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by afishbeck** [comment]
@timothyklemm please review and if possible test this change.

**Review by timothyklemm — CHANGES_REQUESTED** [review]
Preliminary testing was successful. I was able to move some method level transforms to service level. All transforms were processed, as evidenced by seeing the modified request in the log file.

All of the review comments are related to code redundancy. Not sure if old code is still necessary, but if it is it should share code with the new implementation and not duplicate it.

File: esp/services/esdl_svc_engine/esdl_svc_custom.hpp [hunk 1]
```diff
@@ -131,6 +131,8 @@ class CEsdlCustomTransformChoose : public CInterface
 interface IEsdlCustomTransform : extends IInterface
 {
        virtual void processTransform(IEspContext * context, IPropertyTree *tgtcfg, IPropertyTree *tgtctx, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, StringBuffer & request, IPropertyTree * bindingCfg)=0;
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_svc_custom.hpp]: This method could come in handy, I'll keep it for now, but may decide to remove it later.

File: esp/services/esdl_svc_engine/esdl_svc_custom.cpp [hunk 2]
```diff
@@ -310,6 +310,12 @@ CEsdlCustomTransform::CEsdlCustomTransform(IPropertyTree &currentTransform)
 #endif
 }
 
+void CEsdlCustomTransform::processTransform(IEspContext * context, IPropertyTree *theroot, IXpathContext *xpathContext, const char *defaultTarget)
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_svc_custom.cpp]: I've commoned up the code.

File: esp/services/esdl_svc_engine/esdl_svc_custom.cpp [hunk 3]
```diff
@@ -401,3 +407,100 @@ void CEsdlCustomTransform::processTransform(IEspContext * context, IPropertyTree
         ESPLOG(1,"MODIFIED REQUEST: %s", request.str());
     }
 }
+
+void processServiceAndMethodTransforms(IEsdlCustomTransform *serviceCrt, IEsdlCustomTransform *methodCrt, IEspContext * context, IPropertyTree *tgtcfg, IPropertyTree *tgtctx, IEsdlDefService &srvdef, IEsdlDefMethod &mthdef, StringBuffer & request, IPropertyTree * bindingCfg)
```
[the remark this answers is not in the document]
Comment (afishbeck) [reply esp/services/esdl_svc_engine/esdl_svc_custom.cpp]: I've commoned up the code and used an initializer_list.  May switch to a vector later if we need something more flexible.

**Comment by afishbeck** [comment]
@timothyklemm updated and squashed please review.

**Review by timothyklemm — APPROVED** [review]
Reviewed and tested. Service and method level transforms both applied.

**Pull request merged**